### PR TITLE
Test that basic ref field operations don't crash linker

### PR DESCRIPTION
**Pull request by jtschuster**
Adds some tests with basic usage of ref fields to ensure the linker doesn't crash or produce incorrect warnings. Ref structs can't be used as type parameters, so some things like IEnumerable<RS> can't be tested.

Please let me know if I'm missing other cases where this could cause some issues.

### CODE REVIEW

File: test/Mono.Linker.Tests.Cases/Basic/LinkerHandlesRefFields.cs [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+			{
+				this = new RS2 ();
+			}
+		}
+
+		delegate RS2 RsParamRs2Return (RS rs);
+		delegate ref RS2 RefRSParamRefRs2Return (ref RS rs);
+		delegate ref int RsParamRefIntReturn (RS rs);
+
+		public static void Main (string[] args)
```
Comment (vitek-karas) [inline test/Mono.Linker.Tests.Cases/Basic/LinkerHandlesRefFields.cs:50]: I actually don't know if some of these are valid, but still:
- ref struct in a generic parameter
- ref return and parameter on a virtual method?
- Put annotations (DAM/RUC on type) on the ref field - don't care if it works correctly, but should not crash

Comment (jtschuster) [reply test/Mono.Linker.Tests.Cases/Basic/LinkerHandlesRefFields.cs:99]: > ref struct in a generic parameter
Ref structs will generate [CS0306](https://docs.microsoft.com/en-us/dotnet/csharp/misc/cs0306) if used as a type parameter.

> ref return and parameter on a virtual method?
Added IRS interface and Base and Derived classes with virtual methods with ref structs as return, ref return, parameter, and ref parameter.

> Put annotations (DAM/RUC on type) on the ref field
RUC can't be placed on structs and ref structs can't implement interfaces, so I don't think there's a way that they could be annotated with DAM. I've added DAM on a field and DAM and RUC on a method return and parameter.

Comment (vitek-karas) [reply test/Mono.Linker.Tests.Cases/Basic/LinkerHandlesRefFields.cs:99]: Thanks a lot!

File: test/Mono.Linker.Tests.Cases/Basic/LinkerHandlesRefFields.cs [hunk 2]
```diff
@@ -9,6 +9,7 @@
 using System.Text;
 using System.Threading.Tasks;
 using Mono.Linker.Tests.Cases.Expectations.Assertions;
+using Mono.Linker.Tests.Cases.Expectations.Helpers;
 
 namespace Mono.Linker.Tests.Cases.Basic
 {
```
Comment (vitek-karas) [inline test/Mono.Linker.Tests.Cases/Basic/LinkerHandlesRefFields.cs:15]: Nit: Add a comment describing what the test is trying to validate - mainly that it's only meant to validate that linker doesn't crash or do other weird things on ref fields. And that it's not meant as full validation of all functionality around ref fields.

**Review by vitek-karas — APPROVED** [review]
Thanks!

**Pull request merged**